Update MSYS2 base

Hey, this is the future packager for MSYS2. I uploaded a new database recently and reverted it after a few hours due to reports of issues. I even received one issue from a concerned vcpkg user. Due to a bunch of stuff coming together, we don't see a way to make the upgrade path work cleanly for old installations (from an old installer or just not upgraded for a while). Therefore I suggest that vcpkg switch to the current MSYS2 installer which should have the correct keyring. How realistic is this for you?

[notifying @lazka]

The vcpkg log:

```
-- Acquiring MSYS2...
-- Downloading https://sourceforge.net/projects/msys2/files/Base/x86_64/msys2-base-x86_64-20190524.tar.xz/download...
gpg: /etc/pacman.d/gnupg/trustdb.gpg: trustdb created
gpg: no ultimately trusted keys found
gpg: starting migration from earlier GnuPG versions
gpg: porting secret keys from '/etc/pacman.d/gnupg/secring.gpg' to gpg-agent
gpg: migration succeeded
gpg: Generating pacman keyring master key...
gpg: key 69A4A79979944FCF marked as ultimately trusted
gpg: directory '/etc/pacman.d/gnupg/openpgp-revocs.d' created
gpg: revocation certificate stored as '/etc/pacman.d/gnupg/openpgp-revocs.d/5EECEBF3DABCBC4C358437EE69A4A79979944FCF.rev'
gpg: Done
==> Updating trust database...
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
==> Appending keys from msys2.gpg...
==> Locally signing trusted keys in keyring...
  -> Locally signing key D55E7A6D7CE9BA1587C0ACACF40D263ECA25678A...
  -> Locally signing key 123D4D51A1793859C2BE916BBBE514E53E0D0813...
  -> Locally signing key B91BCF3303284BF90CC043CA9F418C233E652008...
  -> Locally signing key 9DD0D4217D75A33B896159E6DA7EF2ABAEEA755C...
==> Importing owner trust values...
gpg: setting ownertrust to 4
gpg: setting ownertrust to 4
gpg: setting ownertrust to 4
gpg: inserting ownertrust of 4
==> Updating trust database...
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   4  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: depth: 1  valid:   4  signed:   3  trust: 0-, 0q, 0n, 4m, 0f, 0u
gpg: depth: 2  valid:   3  signed:   0  trust: 3-, 0q, 0n, 0m, 0f, 0u
:: Synchronizing package databases...
downloading mingw32.db...
downloading mingw32.db.sig...
error: mingw32: key "4A6129F4E4B84AE46ED7F635628F528CF3053E04" is unknown
:: Import PGP key 4096R/87771331B3F1FF5263856A6D974C8BE49078F532, "David Macek <david.macek.0@gmail.com>", created: 2018-01-14? [Y/n] error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update mingw32 (invalid or corrupted database (PGP signature))

downloading mingw64.db...
downloading mingw64.db.sig...
error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update mingw64 (invalid or corrupted database (PGP signature))
downloading msys.db...
downloading msys.db.sig...
error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update msys (invalid or corrupted database (PGP signature))
error: failed to synchronize all databases
error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
checking dependencies...

Packages (2) rebase-4.4.4-1  dash-0.5.10.2-1

Total Removed Size:  1.12 MiB

:: Do you want to remove these packages? [Y/n]
:: Processing package changes...
removing rebase...
removing dash...
error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
:: Synchronizing package databases...
downloading mingw32.db...
downloading mingw32.db.sig...
error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update mingw32 (invalid or corrupted database (PGP signature))
downloading mingw64.db...
downloading mingw64.db.sig...
error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update mingw64 (invalid or corrupted database (PGP signature))
downloading msys.db...
downloading msys.db.sig...
error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
error: failed to update msys (invalid or corrupted database (PGP signature))
error: failed to synchronize all databases
-- Acquiring MSYS2... OK
-- Acquiring MSYS Packages...
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: C:/vcpkg/downloads/tools/msys2/msys64/usr/bin/bash.exe --noprofile --norc -c "pacman -S --noconfirm --needed make automake1.15"
    Working Directory: C:/vcpkg/downloads/tools/msys2
    Error code: 1
    See logs for more information:
     C:\vcpkg\buildtrees\icu\msys-pacman-x64-windows-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_acquire_msys.cmake:127 (vcpkg_execute_required_process)
  ports/icu/portfile.cmake:80 (vcpkg_acquire_msys)
  scripts/ports.cmake:76 (include)
 ```

> The vcpkg log:
> 
> 
> 
> ```
> 
> -- Acquiring MSYS2...
> 
> -- Downloading https://sourceforge.net/projects/msys2/files/Base/x86_64/msys2-base-x86_64-20190524.tar.xz/download...
> 
> gpg: /etc/pacman.d/gnupg/trustdb.gpg: trustdb created
> 
> gpg: no ultimately trusted keys found
> 
> gpg: starting migration from earlier GnuPG versions
> 
> gpg: porting secret keys from '/etc/pacman.d/gnupg/secring.gpg' to gpg-agent
> 
> gpg: migration succeeded
> 
> gpg: Generating pacman keyring master key...
> 
> gpg: key 69A4A79979944FCF marked as ultimately trusted
> 
> gpg: directory '/etc/pacman.d/gnupg/openpgp-revocs.d' created
> 
> gpg: revocation certificate stored as '/etc/pacman.d/gnupg/openpgp-revocs.d/5EECEBF3DABCBC4C358437EE69A4A79979944FCF.rev'
> 
> gpg: Done
> 
> ==> Updating trust database...
> 
> gpg: marginals needed: 3  completes needed: 1  trust model: pgp
> 
> gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
> 
> ==> Appending keys from msys2.gpg...
> 
> ==> Locally signing trusted keys in keyring...
> 
>   -> Locally signing key D55E7A6D7CE9BA1587C0ACACF40D263ECA25678A...
> 
>   -> Locally signing key 123D4D51A1793859C2BE916BBBE514E53E0D0813...
> 
>   -> Locally signing key B91BCF3303284BF90CC043CA9F418C233E652008...
> 
>   -> Locally signing key 9DD0D4217D75A33B896159E6DA7EF2ABAEEA755C...
> 
> ==> Importing owner trust values...
> 
> gpg: setting ownertrust to 4
> 
> gpg: setting ownertrust to 4
> 
> gpg: setting ownertrust to 4
> 
> gpg: inserting ownertrust of 4
> 
> ==> Updating trust database...
> 
> gpg: marginals needed: 3  completes needed: 1  trust model: pgp
> 
> gpg: depth: 0  valid:   1  signed:   4  trust: 0-, 0q, 0n, 0m, 0f, 1u
> 
> gpg: depth: 1  valid:   4  signed:   3  trust: 0-, 0q, 0n, 4m, 0f, 0u
> 
> gpg: depth: 2  valid:   3  signed:   0  trust: 3-, 0q, 0n, 0m, 0f, 0u
> 
> :: Synchronizing package databases...
> 
> downloading mingw32.db...
> 
> downloading mingw32.db.sig...
> 
> error: mingw32: key "4A6129F4E4B84AE46ED7F635628F528CF3053E04" is unknown
> 
> :: Import PGP key 4096R/87771331B3F1FF5263856A6D974C8BE49078F532, "David Macek <david.macek.0@gmail.com>", created: 2018-01-14? [Y/n] error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update mingw32 (invalid or corrupted database (PGP signature))
> 
> 
> 
> downloading mingw64.db...
> 
> downloading mingw64.db.sig...
> 
> error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update mingw64 (invalid or corrupted database (PGP signature))
> 
> downloading msys.db...
> 
> downloading msys.db.sig...
> 
> error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update msys (invalid or corrupted database (PGP signature))
> 
> error: failed to synchronize all databases
> 
> error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> checking dependencies...
> 
> 
> 
> Packages (2) rebase-4.4.4-1  dash-0.5.10.2-1
> 
> 
> 
> Total Removed Size:  1.12 MiB
> 
> 
> 
> :: Do you want to remove these packages? [Y/n]
> 
> :: Processing package changes...
> 
> removing rebase...
> 
> removing dash...
> 
> error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> :: Synchronizing package databases...
> 
> downloading mingw32.db...
> 
> downloading mingw32.db.sig...
> 
> error: mingw32: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update mingw32 (invalid or corrupted database (PGP signature))
> 
> downloading mingw64.db...
> 
> downloading mingw64.db.sig...
> 
> error: mingw64: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update mingw64 (invalid or corrupted database (PGP signature))
> 
> downloading msys.db...
> 
> downloading msys.db.sig...
> 
> error: msys: signature from "David Macek <david.macek.0@gmail.com>" is marginal trust
> 
> error: failed to update msys (invalid or corrupted database (PGP signature))
> 
> error: failed to synchronize all databases
> 
> -- Acquiring MSYS2... OK
> 
> -- Acquiring MSYS Packages...
> 
> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
> 
>     Command failed: C:/vcpkg/downloads/tools/msys2/msys64/usr/bin/bash.exe --noprofile --norc -c "pacman -S --noconfirm --needed make automake1.15"
> 
>     Working Directory: C:/vcpkg/downloads/tools/msys2
> 
>     Error code: 1
> 
>     See logs for more information:
> 
>      C:\vcpkg\buildtrees\icu\msys-pacman-x64-windows-err.log
> 
> 
> 
> Call Stack (most recent call first):
> 
>   scripts/cmake/vcpkg_acquire_msys.cmake:127 (vcpkg_execute_required_process)
> 
>   ports/icu/portfile.cmake:80 (vcpkg_acquire_msys)
> 
>   scripts/ports.cmake:76 (include)
> 
>  ```

--- a/scripts/cmake/vcpkg_acquire_msys.cmake
+++ b/scripts/cmake/vcpkg_acquire_msys.cmake
@@ -81,6 +81,19 @@ function(vcpkg_acquire_msys PATH_TO_ROOT_OUT)
         SHA512 ${HASH}
     )
 
+    # download the new keyring, without it new packages and package updates
+    # might not install
+    vcpkg_download_distfile(KEYRING_PATH
+        URLS http://repo.msys2.org/msys/x86_64/msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz
+        FILENAME msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz
+        SHA512 a5023fd17ccf6364bc6e27c5e63aea25f1fc264a5247cbae4008864c828c38c3e0b4de09ded650e28d2e24e319b5fcf7a9c0da0fa3a8ac81679470fc6bd120c9
+    )
+    vcpkg_download_distfile(KEYRING_SIG_PATH
+        URLS http://repo.msys2.org/msys/x86_64/msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz.sig
+        FILENAME msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz.sig
+        SHA512 c326fefd13f58339afe0d0dc78306aa6ab27cafa8c4d792c2d34aa81fdd1f759d490990ab79daa9664a03a6dfa14ffd2b2ad828bf19a883410112d01f5ed6c4c
+    )
+
     file(REMOVE_RECURSE ${TOOLPATH}/${TOOLSUBPATH})
     file(MAKE_DIRECTORY ${TOOLPATH})
     _execute_process(
@@ -89,6 +102,19 @@ function(vcpkg_acquire_msys PATH_TO_ROOT_OUT)
     )
     _execute_process(
       COMMAND ${PATH_TO_ROOT}/usr/bin/bash.exe --noprofile --norc -c "PATH=/usr/bin;pacman-key --init;pacman-key --populate"
+      WORKING_DIRECTORY ${TOOLPATH}
+    )
+    # install the new keyring
+    _execute_process(
+      COMMAND ${PATH_TO_ROOT}/usr/bin/bash.exe --noprofile --norc -c "PATH=/usr/bin;pacman-key --verify ${KEYRING_SIG_PATH}"
+      WORKING_DIRECTORY ${TOOLPATH}
+      RESULT_VARIABLE _vam_error_code
+    )
+    if(_vam_error_code)
+      message(FATAL_ERROR "Cannot verify MSYS2 keyring.")
+    endif()
+    _execute_process(
+      COMMAND ${PATH_TO_ROOT}/usr/bin/bash.exe --noprofile --norc -c "PATH=/usr/bin;pacman -U ${KEYRING_PATH} --noconfirm"
       WORKING_DIRECTORY ${TOOLPATH}
     )
     # we have to kill all GnuPG daemons otherwise bash would potentially not be

